src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingCommandMessageWireProtocol.cs

```diff
@@ -266,7 +278,16 @@ private TCommandResult ProcessResponse(ConnectionId connectionId, CommandMessage
                     _session.AdvanceOperationTime(operationTime.AsBsonTimestamp);
                 }
 
-                if (!rawDocument.GetValue("ok", false).ToBoolean())
+                if (rawDocument.GetValue("ok", false).ToBoolean())
+                {
+                    if (rawDocument.Contains("recoveryToken"))
+                    {
+                        var materializedDocument = rawDocument.Materialize(binaryReaderSettings);
+                        var recoveryToken = materializedDocument["recoveryToken"].AsBsonDocument;
+                        _session.CurrentTransaction.RecoveryToken = recoveryToken;
+                    }
+                }
```

Comment:
Note: for a moment I was wondering whether we need to check whether CurrentTransaction is null, but I think we don't need to because if we see a recovery token we can assume we must be in a transaction because the server won't return a recovery token if we are not in a transaction.